airflow/executors/kubernetes_executor.py

```diff
@@ -171,16 +204,21 @@ def _run(
 
         return last_resource_version
 
-    def process_error(self, event: Any) -> str:
+    def process_error(
+        self,
+        event: Any,
+        kube_client: client.CoreV1Api,
+    ) -> str:
         """Process error response"""
         self.log.error('Encountered Error response from k8s list namespaced pod stream => %s', event)
         raw_object = event['raw_object']
         if raw_object['code'] == 410:
             self.log.info(
-                'Kubernetes resource version is too old, must reset to 0 => %s', (raw_object['message'],)
+                'Kubernetes resource version is too old, '
+                'relisting pods to get the latest version. Error => %s',
+                (raw_object['message'],),
             )
-            # Return resource version 0
-            return '0'
+            return get_resource_version(kube_client, self.namespace)
```

Comment:
I wonder if this would be cleaner if we return `None` here and let the next `_run` relist at the start? It would also keep the list specifics in one spot vs both here and in `_make_kube_watcher`. Thoughts?

Reply:
If we return `None` here the code would be ugly, because then, we have to check the returned value of `_run` method in `run` method. If `_run` returns None, we call get_resource_version and then set self.resource_version.

Still the same behaviour but ugly. This is what I think it would look like:

```python
    def run(self) -> None:
        """Performs watching"""
        kube_client: client.CoreV1Api = get_kube_client()
        if not self.scheduler_job_id:
            raise AirflowException(NOT_STARTED_MESSAGE)
        while True:
            try:
                self.resource_version = self._run(
                    kube_client, self.resource_version, self.scheduler_job_id, self.kube_config
                )
                if self.resource_version is None:
                    self.resource_version = get_resource_version(kube_client, self.namespace)
            except ReadTimeoutError:
                self.log.warning(
                    "There was a timeout error accessing the Kube API. Retrying request.", exc_info=True
                )
                time.sleep(1)
            except Exception:
                self.log.exception('Unknown error in KubernetesJobWatcher. Failing')
                raise
            else:
                self.log.warning(
                    'Watch died gracefully, starting back up with: last resource_version: %s',
                    self.resource_version,
                )
```
Which gives the same result, but all these was solved by calling the get_resource_version on error

Reply:
> if the version is too old, and we start at a _later_ version, might that not mean we've missed some events, so re-listing is the "correct" behaviur

Yeah, this was always an issue and continues to be, this is just the right way to start watching again. What is missing is reconciling our state vs the list to figure out the events we missed, but that is a bigger fish imo.